### PR TITLE
Add Generate GO genesets processor

### PR DESCRIPTION
**Pull request by jkokosar**

### CODE REVIEW

File: resolwe_bio/tools/go_genesets.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+from resolwe_runtime_utils import export  # pylint: disable=import-error
+from utils import gzopen  # pylint: disable=import-error
+
+
+def parse_arguments():
+    """Parse command line arguments."""
+    parser = argparse.ArgumentParser(description="Parse Diff Exp output files")
+    parser.add_argument('gaf_file', help="GAF file.")
+    parser.add_argument('--source', help="GeneID source.")
```
Comment (dblenkus) [inline resolwe_bio/tools/go_genesets.py]: This field is required in `'upload-geneset` process, so it cannot be optional here.

Comment (jkokosar) [inline resolwe_bio/tools/go_genesets.py]: Fixed!

File: resolwe_bio/processes/support_processors/go_genesets.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# ====================
+# Generate GO genesets
+# ====================
+---
+
+- slug: go-genesets
+  name: Generate GO genesets
```
Comment (mstajdohar) [inline resolwe_bio/processes/support_processors/go_genesets.yml]: Geneset should be written with space: "gene set". Fix this everywhere. Here I suggest to change the name to "Import gene sets: GO terms".

File: resolwe_bio/processes/support_processors/go_genesets.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  data_name: "Generate genesets ({{gaf.gaf.file}})"
```
Comment (mstajdohar) [inline resolwe_bio/processes/support_processors/go_genesets.yml]: Rename to Import gene sets from {{gaf.gaf.file}}

File: resolwe_bio/tools/geneset2storage.py [hunk 4]
```diff
@@ -19,7 +21,12 @@ def main():
     with open(args.geneset_file) as infile:
         geneset = [gene.strip() for gene in infile]
         data = {'geneset_json': geneset}
```
Comment (mstajdohar) [inline resolwe_bio/tools/geneset2storage.py]: Rename the inner `geneset_json` to `genes` and fix generators, endpoints, queries... Discuss with the product team.

File: resolwe_bio/tools/geneset2storage.py [hunk 5]
```diff
@@ -19,7 +21,12 @@ def main():
...
-        print(json.dumps(data, separators=(',', ':'), allow_nan=False))
+
+        if args.output:
```
Comment (mstajdohar) [inline resolwe_bio/tools/geneset2storage.py]: Support just the solution that works for large number of genes.

File: resolwe_bio/tools/go_genesets.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+from collections import defaultdict
+
+from resolwe_runtime_utils import export  # pylint: disable=import-error
+from utils import gzopen  # pylint: disable=import-error
+
+
+def parse_arguments():
+    """Parse command line arguments."""
+    parser = argparse.ArgumentParser(description="Parse Diff Exp output files")
```
Comment (mstajdohar) [inline resolwe_bio/tools/go_genesets.py]: Wrong description.

File: resolwe_bio/processes/import_data/differential_exp.yml [hunk 7]
```diff
@@ -84,6 +84,18 @@
           value: mirna20
         - label: mirna30 (Affy)
           value: mirna30
+    - name: case
```
Comment (mstajdohar) [inline resolwe_bio/processes/import_data/differential_exp.yml]: Maybe check in the beginning of the process algorithm that if cases are given, controls should be given too.

File: docs/CHANGELOG.rst [hunk 8]
```diff
@@ -24,13 +19,16 @@ Changed
...
+* Refactor ``upload-diffexp``processor
 
 Added
 -----
 * Process definition documentation style and text improvements
+* Add tests to ensure generators produce the same results
 * Upload Gene sets (``data:geneset``)
 * Add ``generate_geneset`` django-admin command
 * Add ``generate_diffexpr_deseq`` django-admin command
+* Add 'Generate GO genesets' processor
```
Comment (mstajdohar) [inline docs/CHANGELOG.rst]: Change genesets to gene sets

File: resolwe_bio/processes/support_processors/go_genesets.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
+# ====================
+# Generate GO genesets
```
Comment (mstajdohar) [inline resolwe_bio/processes/support_processors/go_genesets.yml]: Change genesets to gene sets

File: resolwe_bio/processes/support_processors/go_genesets.yml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  input:
+    - name: gaf
+      label: Reference GAF file
+      type: data:gaf:2:0
+    - name: source
+      label: Gene ID source
+      type: basic:string
+  output:
+    - name: num_genesets
+      label: Number of genesets generated
```
Comment (mstajdohar) [inline resolwe_bio/processes/support_processors/go_genesets.yml]: Change genesets to gene sets

File: resolwe_bio/tools/go_genesets.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env python2
+"""Parse GAF file to genesets."""
```
Comment (mstajdohar) [inline resolwe_bio/tools/go_genesets.py]: Change genesets to gene sets

File: resolwe_bio/tools/go_genesets.py (hunk 6 above)
Comment (mstajdohar) [inline resolwe_bio/tools/go_genesets.py]: Still not addressed.

**Pull request merged**